lib/OpenQA/Task/Utils.pm

```diff
@@ -48,5 +51,53 @@ sub finish_job_if_disk_usage_below_percentage (%args) {
     return 1;
 }
 
+# define a signal handler for retrying the job after receiving a signal
+my @GUARDS;
+my $STOP_AFTER_LOCK_ACQUISITION = 0;
+my $CONCLUDE_ON_SIGNALS = 0;
+sub _handle_signal ($job, $signal) {
+    # do nothing if the job is supposed to be concluded despite receiving a signal at this point
+    return $job->note(signal_handler => "Received signal $signal, concluding") if $CONCLUDE_ON_SIGNALS;
+
+    # schedule a retry before stopping the job's execution prematurely
+    $job->note(signal_handler => "Received signal $signal, scheduling retry and releasing locks");
+    $job->retry;
+
+    if (@GUARDS) {
+        # release locks and exit if lock acquisition has been concluded at this point
+        undef $$_ for @GUARDS;
+        exit;
```

Comment:
Are you sure this is necessary? I'd expect `exit` to allow cleanup code in scope guards to run.

Reply:
Or is it the timing of destructors resulting in the database connection already being closed when the guard is garbage collected? 🤔

Reply:
If every guard really needs to be registered in the future we should probably have a helper that combines creating the guard and registering it.

Reply:
You're right. I must have done something wrong on my first attempt (e.g. sending SIGKILL instead of SIGTERM). That simplifies things greatly. I'll update the PR tomorrow.

Reply:
Ok, I've just pushed it now. That greatly simplifies things :-)